modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+                        return false;
+                    }
+
+                    // Read the actual key into a direct buffer.
+                    it.key(directBuffer);
+
+                    // To understand this condition please read that huge comment above.
+                    if (timestamp == null || !directBuffer.equals(seekKeyBufSliceWithoutTimestamp)) {
+                        // Copy actual row id into a "seekKeyBuf" buffer.
+                        GridUnsafe.copyMemory(
+                                null, GridUnsafe.bufferAddress(directBuffer) + ROW_ID_OFFSET,
+                                seekKeyBuf.array(), GridUnsafe.BYTE_ARR_OFF + ROW_ID_OFFSET,
+                                igniteRowIdSize
+                        );
+
+                        // Perform additional "seek" if timestamp is not null. Motivation for it is described in comments above.
+                        if (timestamp != null) {
+                            // At this point, "seekKeyBuf" has row id that exists in partition.
+                            it.seek(seekKeyBuf.array());
+
+                            // Iterator may not be valid if that row was created after required timestamp.
+                            if (invalid(it)) {
+                                return false;
+                            }
+
+                            // Or iterator may still be valid even if there's no version for required timestamp. In this case row id
+                            // itself will be different and must check it.
+                            it.key(directBuffer);
+
+                            if (!directBuffer.equals(seekKeyBufSliceWithoutTimestamp)) {
+                                found = false;
+                            }
+                        }
+                    }
+
+                    // This one might look tricky. We finished processing next row. There are three options:
+                    //  - "found" flag is false - there's no fitting version of the row. We'll continue to next iteration;
+                    //  - value is empty, we found a tombstone. We'll continue to next iteration as well;
+                    //  - value is not empty and everything's good. We'll cache it and return from method.
+                    // In all three cases we need to prepare the value of "seekKeyBuf" so that it has not-yet-scanned row id in it.
+                    // the only valid way to do so is to treat row id payload as one big unsigned integer in Big Endian and increment it.
+                    // It's important to note that increment may overflow. In this case "carry flag" will go into incrementing partition id.
+                    // This is fine for three reasons:
+                    //  - iterator has an upper bound, following "seek" will result in invalid iterator state.
+                    //  - partition id iself cannot be overflown, because it's limited with a constant less than 0xFFFF. It's something like
```

Comment:
I suggest adding the following assertion (on the mentioned constant) to the static initializer of this class:

```
assert MAX_PARTITION_ID < 0xFFFF;
```

To make sure that this invariant is kept in the future.

Reply:
Really? There's no point in it, it's a core invariant of the system

Reply:
I feel really unconfortable when a core invariant is not verified with some test or production code :) Because it is so easy to break an invariant that is not verified automatically.

If it is already verified, then ok.

Reply:
We have a specific constant, INDEX_PARTITION which is 0xFFFF. Regular partitions can't reach it, I see no point in testing it. It's not obvious, I get it, but this is a part of the design